### PR TITLE
remove QP's dependency on config

### PR DESCRIPTION
**Pull request by shyamjesal**
- Replace config file with environment variables
  - split env variables in 2 files: network and xdt 
  - Added struct tags in `utils/common.go` for default config params.
  - `utils/common.go` now parses environment variables using [envconfig](github.com/kelseyhightower/envconfig) instead of config file
  - Added environment variables in Makefile
  - Created `user-functions/gx/gx.yaml` and `user-functions/fx/fx.yaml` for knative deployment

- Removed redundent`ChunkSize` in `proto/crossXDT` `proto/downXDT`
- Manually append headers in http proxy as extracting metadata from `r.Context()` does not work
- Removed Payload un-marshalling at `sdk/golang/destination.go` as the metadata is send using http headers.
- Removed global config variable in `utils/common.go`
  - removed global config and handler variables in `sdk/golang/destination.go`
- Changed default routing mode to "CutThrough" 
- changed h2c ports to 50051 in fx.yaml and gx.yaml

### CODE REVIEW

**Review by ustiugov — CHANGES_REQUESTED** [review]
need to be re-done using env variables instead of headers. btw, you may want to keep you LoadConfig function but in the knative mode, it should parse the env variables. **Do not parse them in many places in the code.**

File: .github/workflows/xdt_tests.yml [hunk 1]
```diff
@@ -93,8 +93,26 @@ jobs:
       - name: Check out code into the Go module directory
         uses: actions/checkout@v2
       - name: Run benchmarks
+        env:
+          CHUNKSIZEINBYTES: 65536
```
Comment (ustiugov) [inline .github/workflows/xdt_tests.yml]: it's not readable, please use underscores to separate words

File: user-functions/.env [hunk 2]
```diff
@@ -0,0 +1,18 @@
+CHUNKSIZEINBYTES=65536
```
Comment (ustiugov) [inline user-functions/.env]: what is this file? why don't makefiles use it?

File: user-functions/fx/fx.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  name: fx
+  namespace: default
+spec:
+  template:
+    spec:
+      containers:
+        - image: docker.io/vhiveease/fx:latest
+          imagePullPolicy: Always
+          env:
+            - name: CHUNKSIZEINBYTES
```
Comment (ustiugov) [inline user-functions/fx/fx.yaml]: in the knative setup, hostnames and ports should not be configurable, they should have the default values.

please clear the redundant parameters

File: utils/common.go [hunk 4]
```diff
@@ -23,90 +23,50 @@
...
-		if err != nil {
-			log.Errorf("transport: Error closing the config file")
-		}
-	}()
-
-	jsonByteValue, err := ioutil.ReadAll(jsonFile)
-	if err != nil {
-		log.Fatal(err)
-	}
+var LoadConfig = ReadConfig()
```
Comment (ustiugov) [inline utils/common.go]: instead of a global variable, every component should call ReadConfig(). Globals are bad style

**Review by ustiugov — CHANGES_REQUESTED** [review]
- [x] use underscores for env variables for readability
- [x] why routing is still a runtime argument? it should be via env vars
- [x] in the knative setup, hostnames and ports should not be configurable, they should have the default values. please remove redundant parameters
- [x] remove the global LoadConfig variable

File: user-functions/.env (hunk 2 above)
Comment (shyamjesal) [reply user-functions/.env]: this file is for loading env variables for docker compose run

File: queue-proxy/sQP/sQP.go [hunk 5]
```diff
@@ -61,11 +61,13 @@ func (s upXDTServer) SendData(srv upXDT.StreamData_SendDataServer) error {
...
 	var channel chan []byte
 	var totalChunks int64
+	headers, _ := metadata.FromIncomingContext(srv.Context())
+	log.Infof("sQP: using %s routing", headers["routing"][0])
 	for {
 		chunk, err := srv.Recv()
 		if err == io.EOF {
 			log.Infof("sQP: %d chunks received", chunkCount)
-			if config.Routing == utils.STORE_FORWARD {
+			if headers["routing"][0] == utils.STORE_FORWARD {
```
Comment (ustiugov) [inline queue-proxy/sQP/sQP.go:70]: this is what I don't get. why is routing in the header? it should be upon startup with an env variable, right?

File: sdk/golang/destination.go [hunk 6]
```diff
@@ -39,50 +40,47 @@ import (
...
-	if err := json.Unmarshal(in.XDTJSON, &xdtPayload); err != nil {
-		log.Error("DST: XDTFnCall", err)
-		return &downXDT.Empty{}, err
-	}
+	headers, ok := metadata.FromIncomingContext(ctx)
 
-	key := xdtPayload.Key
+	if ok && headers["is_xdt"][0] == "true" {
+		key := headers["key"][0]
+		log.Infof("DST: using %s routing", headers["routing"][0])
```
Comment (ustiugov) [inline sdk/golang/destination.go:57]: shouldn't be in headers

File: sdk/python/source.py [hunk 7]
```diff
@@ -53,32 +53,32 @@ def splitPayload(xdtPayload):
...
 
+    sQPAddr = config["SQPServerHostname"]+config["SQPServerPort"]
     key, payloadData, xdtPayload = splitPayload(xdtPayload)
     serialisedPayload = xdtPayload.tobytes()
 
     metadata = (
         ('is_xdt', 'true'),
         ('key', key),
         ('sqp_addr', sQPAddr),
+        ('routing', config['Routing']),
```
Comment (ustiugov) [inline sdk/python/source.py:66]: shouldn't be here

File: sdk/python/test.py [hunk 8]
```diff
@@ -39,23 +39,27 @@ def test_splitPayload(self):
         assert data == b'0123456789'
 
     def test_Push_data(self):
-        PushData(key='secret', payload=b'01234567890',
+        metadata = (
+            ('is_xdt', 'true'),
+            ('key', 'secret'),
+            ('sqp_addr', config['SQPServerHostname']+config['SQPServerPort']),
+            ('routing', config['Routing']),
```
Comment (ustiugov) [inline sdk/python/test.py:46]: shouldn't be in a header

File: tests/integration_test.go [hunk 9]
```diff
@@ -96,27 +99,33 @@ func knativeQP(config utils.Config) {
...
 			defer h.ServeHTTP(w, r)
 
-			isXDT := r.Header.Get("is_xdt")
-			if isXDT == "true" {
+			if r.Header.Get("is_xdt") == "true" {
+				httpMetadata := map[string]string{
+					"is_xdt":   r.Header.Get("is_xdt"),
+					"key":      r.Header.Get("key"),
+					"sqp_addr": r.Header.Get("sqp_addr"),
+					"routing":  r.Header.Get("routing"),
```
Comment (ustiugov) [inline tests/integration_test.go:113]: shouldn't be in a header

File: user-functions/dQP/dQP.go [hunk 10]
```diff
@@ -82,18 +83,26 @@ func main() {
...
-				log.Infof("pulling from sQP using key %s addr %s", r.Header.Get("key"), r.Header.Get("sQPAddr"))
-				err := dQP.PullDataFromSrcQP(r.Context(), r.Header.Get("key"), r.Header.Get("sQPAddr"), config.ChunkSizeInBytes)
-				if err != nil {
-					log.Errorf("Proxy: unable to pull data from sQP: %v", err)
+			if r.Header.Get("is_xdt") == "true" {
+				httpMetadata := map[string]string{
+					"is_xdt":   r.Header.Get("is_xdt"),
+					"key":      r.Header.Get("key"),
+					"sqp_addr": r.Header.Get("sqp_addr"),
+					"routing":  r.Header.Get("routing"),
```
Comment (ustiugov) [inline user-functions/dQP/dQP.go:91]: shouldn't be in a header

File: Makefile [hunk 11]
```diff
@@ -29,6 +29,25 @@ ROUTING_MODES:=Store&Forward CutThrough
 GO_TEST_FLAGS:=-race -v -cover
 SDK_GO_FILES:=./source.go ./destination.go
 
+.EXPORT_ALL_VARIABLES:
```
Comment (ustiugov) [inline Makefile:32]: is it possible to import it from the .env file that you use for docker compose? 2 sets of vars is hard to maintain

File: user-functions/fx/fx.yaml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+              value: "1000"
+            - name: RPC_TIMEOUT_MAX_BACK_OFF
+              value: "60000"
+            - name: RPC_TIMEOUT_DURATION
+              value: "1"
+            - name: RPC_RETRY_DELAY
+              value: "10"
+          ports:
+            - name: h2c
+              containerPort: 50010
```
Comment (ustiugov) [inline user-functions/fx/fx.yaml]: please use the same 50051 port in knative in all functions

File: user-functions/gx/gx.yaml [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+              value: "1000"
+            - name: RPC_TIMEOUT_MAX_BACK_OFF
+              value: "60000"
+            - name: RPC_TIMEOUT_DURATION
+              value: "1"
+            - name: RPC_RETRY_DELAY
+              value: "10"
+          ports:
+            - name: h2c
+              containerPort: 50007
```
Comment (ustiugov) [inline user-functions/gx/gx.yaml]: 50051 please

**Review by ustiugov — CHANGES_REQUESTED** [review]

**Review by ustiugov — APPROVED** [review]

**Pull request merged**